README.md

# [Alto en X](https://nebil.github.io/alto-en-x)

Un sitio bajo en kilobytes y casi sin dependencias (únicamente _[normalize.css](
https://necolas.github.io/normalize.css)_) que permite crear sellos _ad libitum_.  
La idea es poder generar etiquetados —como los que aparecen en los alimentos—
con el uso de estándares web.

## Requisitos

Para funcionar de forma correcta, este sitio necesita un navegador relativamente moderno.  
Sin embargo, cualquier _web browser_ que entienda [SVG 1.1](http://caniuse.com/#feat=svg)
debería cumplir con esta tarea.  
Por ejemplo, la última versión de Mozilla [Firefox](
https://www.mozilla.org/firefox/new) o Google [Chrome](
https://www.google.com/chrome) será más que suficiente.

[:fox_face:](https://www.mozilla.org/en-US/firefox/products) 
_Mucho gusto. Así es, con un poco de imaginación, los zorros podemos hablar._  
     
_Por supuesto, estás invitado a hacer clic en mi seguro, libre y fantástico pelaje._

## Uso alternativo

Es posible, además, crear sellos con el
_[query string](https://tools.ietf.org/html/rfc3986#section-3.4)_;
en efecto, basta con agregar `?input=<texto>` al final del URL.  
Por ejemplo, `https://nebil.github.io/alto-en-x/?input=ejemplo`
nos generará un sello que dirá “[alto&nbsp;en&nbsp;ejemplo](
https://nebil.github.io/alto-en-x/?input=ejemplo)”.

## Diseño

### De la geometría

El sello está escrito en [SVG](https://en.wikipedia.org/wiki/Scalable_Vector_Graphics),
con el propósito de representar un octágono[\*](#one) regular de **múltiples** bordes.  
Lamentablemente, **no es posible** definir este tipo de bordes a partir de
la [actual especificación](https://www.w3.org/TR/SVG) de SVG,  
pero es altamente probable que sí estén disponibles
en la [siguiente versión](https://www.w3.org/TR/SVG2/painting.html) :grinning:
—que quizá [nunca veremos](https://css-tricks.com/svg-2-conundrum). :roll_eyes:  
Por lo tanto, el _workaround_ consiste en elaborar cuatro `<polygon>` superpuestos,
con bordes de distinto ancho.

### De la tipografía

El manual elaborado por el Ministerio de Salud establece que

> La familia Arial es la tipografía utilizada en la iconografía,  
> específicamente su presentación _bold_ para la mayor legibilidad de los textos.

Sin embargo, debemos recordar que Arial es una familia
[comercial](https://www.fonts.com/font/monotype/arial). :moneybag:  
Algunos navegadores cuentan con [ella](https://en.wikipedia.org/wiki/Arial),
otros con [Helvetica](https://en.wikipedia.org/wiki/Helvetica),
y las personas [libres](https://www.stallman.org) disfrutan de
[Liberation Sans](https://en.wikipedia.org/wiki/Liberation_fonts).  
Pero, lamentablemente, existen dispositivos que no conocen ninguna de ellas.
¿Qué podemos hacer ante esto?

> _Introducing Arimo: one family to rule them all._

O en otras palabras, [Arimo](https://fonts.google.com/specimen/Arimo)
—que es idéntico a Liberation Sans— es el nuevo _typeface_ estándar,  
con el objetivo de ofrecer un mismo sello, **visualmente consistente**,
en todos los navegadores[\*\*](#two).  
Además, por temas de seguridad,
las imágenes generadas desde SVG no aceptan tipofaces externas.  
Para resolver este problema, Arimo fue transformada a
[base64](https://en.wikipedia.org/wiki/Base64),
para luego ser insertada en el octágono.

## Créditos

Desde acá, aprovecharé de agradecer

- al [Ministerio de Salud](http://www.minsal.cl) por el diseño de los etiquetados
- a Steve Matteson por el refrescante diseño de
[Arimo](https://fonts.google.com/specimen/Arimo)
- a [Google Fonts](https://fonts.google.com) por ofrecer dos elegantes tipofaces
  - [Fira Sans](https://fonts.google.com/specimen/Fira+Sans), obra de Carrois Apostrophe
  - [Roboto](https://fonts.google.com/specimen/Roboto), obra de Christian Robertson
- a [@necolas](https://github.com/necolas) por _normalize.css_
- a [@mrpatiwi](https://github.com/mrpatiwi) por la idea.

## Licencia

Copyright © 2017, Nebil Kawas García  
El código de este repositorio está bajo el [Mozilla Public License v2.0](
https://www.mozilla.org/MPL/2.0/).

---

<a name='one'>\*</a>
también conocido como _octógono_, pero mantendremos las raíces helénicas.  
<a name='two'>\*\*</a>
donde “ todos los navegadores” quiere decir “ojalá, todos los navegadores”.


CHANGELOG.md

# _Changelog_

Todos los cambios relevantes en este repositorio serán descritos en este archivo.  
Este proyecto cumplirá con _[semantic versioning](http://semver.org)_
—o, por lo menos, lo intentará. :grinning:

-----------------------------------------------------
### [0.3.1](../../compare/0.3.0...0.3.1) – 2017-05-21

#### :rabbit: _¿Qué hay de nuevo, viejo?_

Este nuevo _release_ trae algunos cambios menores,
que **no modifican** la funcionalidad principal.  
Entre ellos, podemos destacar

- un ajuste del color del _footer_ para lograr una mejor amalgama cromática
- otro _upgrade_ [[?](https://github.com/necolas/normalize.css/issues/664)]
de la librería _[normalize.css](https://necolas.github.io/normalize.css)_
(6.0.0 [⟶](https://github.com/necolas/normalize.css/compare/6.0.0...7.0.0) 7.0.0)
- y un _refactor_ de JavaScript que separa el código en dos módulos independientes.

----------------------------------------------------
## [0.3.0](../../compare/0.2.1...0.3.0) – 2017-04-30

#### :rabbit: _¿Qué hay de nuevo, viejo?_

Este flamante _release_ viene con algunas mejoras,
que **sí modifican** la funcionalidad principal.  
Entre ellas, podemos destacar

- un ajuste automático del tamaño del texto,
considerando el largo de este mismo
- la posibilidad de generar sellos utilizando el _[query string](
README.md#uso-alternativo)_ del URL
- un versátil catálogo heptacromático para el fondo del sitio web
- un poco de _metadata_ para ofrecer _[summary cards](
https://dev.twitter.com/cards/types/summary)_ en Twitter
- un discreto _upgrade_ de la librería _[normalize.css](
https://necolas.github.io/normalize.css)_
(5.0.0 [⟶](https://github.com/necolas/normalize.css/compare/5.0.0...6.0.0) 6.0.0)
- y, finalmente, un imperceptible _refactor_ de JavaScript.

-----------------------------------------------------
### [0.2.1](../../compare/0.2.0...0.2.1) – 2017-03-22

#### :rabbit: _¿Qué hay de nuevo, viejo?_

Este nuevo _release_ trae algunos cambios menores,
que **no modifican** la funcionalidad principal.  
Entre ellos, podemos destacar

- un ajuste en el nombre del archivo descargado,
para los sellos formados por dos o más palabras
- el cambio de [TrueType](https://en.wikipedia.org/wiki/TrueType_fonts)
a [WOFF](https://en.wikipedia.org/wiki/Web_Open_Font_Format),
reduciendo el tamaño del _font_ a menos de la mitad
- un par de mejoras sutiles en el _user experience_
- y un pequeño _refactor_ de JavaScript.

----------------------------------------------------
## [0.2.0](../../compare/0.1.1...0.2.0) – 2017-03-12

#### :rabbit: _¿Qué hay de nuevo, viejo?_

Este flamante _release_ viene con algunas mejoras,
que **sí modifican** la funcionalidad principal.  
Entre ellas, podemos destacar

- la posibilidad de elaborar sellos que requieran un texto de dos líneas
- la incorporación de Arimo Bold como nuevo _typeface_ para los sellos
- un enlace directo hacia el _changelog_ del repositorio
- unas breves notas sobre las [decisiones de diseño](README.md#diseño)
- un poco de _metadata_ acerca del sitio web
- y, finalmente, un [favícono](https://en.wikipedia.org/wiki/Favicon) minimalista.

-----------------------------------------------------
### [0.1.1](../../compare/0.1.0...0.1.1) – 2017-02-22

#### :rabbit: _¿Qué hay de nuevo, viejo?_

Este nuevo _release_ trae algunos cambios menores,
que **no modifican** la funcionalidad principal.  
Entre ellos, podemos destacar

- un enlace al repositorio ubicado en GitHub
- una superflua aclaración con relación al [Minsal](http://www.minsal.cl)
- la inserción de un ejemplo al final del párrafo principal
- algunos minuciosos ajustes tipográficos para mi bienestar mental
- dos nuevos archivos: uno para el _script_, otro para los estilos
- y, finalmente, un _refactor_ —casi imperceptible— de CSS y de JavaScript.

---------------------
## 0.1.0 – 2017-02-02

:sparkles: _Voilà!_  
Aquí tenemos el flamante primer _release_ de _Alto en X_.  
El propósito es bastante simple:
construir un generador de etiquetados usando tecnologías web.

### _Features_

Esta pequeña primera versión permite

- generar, de forma simple, sellos personalizables
- descargar el resultado en dos posibles formatos: [PNG] y [SVG].

[/]:# (Referencias implícitas)

[png]: https://en.wikipedia.org/wiki/Portable_Network_Graphics
[svg]: https://en.wikipedia.org/wiki/Scalable_Vector_Graphics


script.js
/*
Copyright (c) 2017, Nebil Kawas García
This source code is subject to the terms of the Mozilla Public License.
You can obtain a copy of the MPL at <https://www.mozilla.org/MPL/2.0/>.
*/

(function() {
function setText() {
    var trimmedText = input.value.trim();
    var altoEnText = document.getElementById('alto-en');
    var upperText = document.getElementById('upper');
    var lowerText = document.getElementById('lower');
    var minsalText = document.getElementById('minsal');
    var replacement = trimmedText.toUpperCase() || ' ';
    upperText.textContent = replacement;

    var hasWhitespace = replacement.indexOf(' ') !== -1;
    if (hasWhitespace && upperText.getComputedTextLength() > 108) {
        // First, set the text contents.
        var wordList = replacement.split(/\s+/);
        upperText.textContent = wordList[0];
        lowerText.textContent = wordList.slice(1).join(' ');
        // REVIEW: The number of words in each line should be picked
        // ======= in a more aesthetic way, by automatically finding
        //         an arrangement that provides an equipoise between
        //         the length of both lines.

        // Then, adjust all the positions.
        altoEnText.setAttribute('y', 50);
        upperText.setAttribute('dy', 22);
        lowerText.setAttribute('dy', 22);
        minsalText.setAttribute('dy', 21);
    } else {
        // Otherwise, reset everything.
        lowerText.textContent = '';
        altoEnText.setAttribute('y', 60);
        upperText.setAttribute('dy', 25);
        lowerText.setAttribute('dy', 00);
        minsalText.setAttribute('dy', 30);
    }

    function updateFontSize(newSize) {
        upperText.style.fontSize = newSize;
        lowerText.style.fontSize = newSize;
        upperLength = upperText.getComputedTextLength();
        lowerLength = lowerText.getComputedTextLength();
    }

    var upperLength = upperText.getComputedTextLength();
    var lowerLength = lowerText.getComputedTextLength();
    // Harmonize the text size, depending on its length.
    if (upperLength <= 100 && lowerLength <= 100)
        updateFontSize('20px');
    if (upperLength > 100 || lowerLength > 100)
        updateFontSize('19px');
    if (upperLength > 118 || lowerLength > 118)
        updateFontSize('18px');

    // Adapted from: https://developer.mozilla.org/en-US/docs/Web/API/
    // ======= =====     Canvas_API/Drawing_DOM_objects_into_a_canvas/
    var container = document.getElementById('container');
    var svgString = new XMLSerializer().serializeToString(container);
    var imageBlob = new Blob([svgString], { type: 'image/svg+xml' });

    var canvas = document.getElementById('hidden-canvas');
    var domUrl = window.URL || window.webkitURL || window;
    var svg = domUrl.createObjectURL(imageBlob);
    var img = new Image();
    img.src = svg;

    img.onload = function() {
        canvas.getContext('2d').drawImage(img, 0, 0);
        var caterpillarName = trimmedText.toLowerCase()
                                         .split(/\s+/)
                                         .join('-');
        var downloadFilename = 'alto-en-' + caterpillarName;

        // Offer a PNG version.
        var png = canvas.toDataURL('image/png');
        var pngButton = document.getElementById('to-png');
        pngButton.download = downloadFilename + '.png';
        pngButton.href = png;

        // Offer an SVG version.
        var svgButton = document.getElementById('to-svg');
        svgButton.download = downloadFilename + '.svg';
        svgButton.href = svg;
    };
}

function fetchQueryValue(key) {
    var regex = '[?&]' + key + '=[^&]*';
    var match = RegExp(regex).exec(window.location.search);
    return match && decodeURIComponent(match[0].split('=')[1]);
}

var input = document.getElementById('text-input');
input.addEventListener('input', setText);

if (value = fetchQueryValue('input')) {
    input.value = value;
    setText();
}

document.getElementById('upper').addEventListener('click', function() {
    input.focus();
});
}());
